pkg/key/soft_key.go

```diff
@@ -0,0 +1,353 @@
+func LoadSoft(networkID uint32, keyPath string) (*SoftKey, error) {
+	kb, err := os.ReadFile(keyPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// in case, it's already encoded
+	k, err := NewSoft(networkID, WithPrivateKeyEncoded(string(kb)))
+	if err == nil {
+		return k, nil
+	}
+
+	r := bufio.NewReader(bytes.NewBuffer(kb))
+	buf := make([]byte, privKeySize)
+	n, err := readASCII(buf, r)
+	if err != nil {
+		return nil, err
+	}
+	if n != len(buf) {
+		return nil, ErrInvalidPrivateKeyLen
+	}
+	if err := checkKeyFileEnd(r); err != nil {
+		return nil, err
+	}
+
+	skBytes, err := hex.DecodeString(string(buf))
+	if err != nil {
+		return nil, err
+	}
+	rpk, err := keyFactory.ToPrivateKey(skBytes)
+	if err != nil {
+		return nil, err
+	}
+	privKey, ok := rpk.(*crypto.PrivateKeySECP256K1R)
+	if !ok {
+		return nil, ErrInvalidType
+	}
+
+	return NewSoft(networkID, WithPrivateKey(privKey))
+}
+
+// readASCII reads into 'buf', stopping when the buffer is full or
+// when a non-printable control character is encountered.
+func readASCII(buf []byte, r io.ByteReader) (n int, err error) {
+	for ; n < len(buf); n++ {
+		buf[n], err = r.ReadByte()
+		switch {
+		case errors.Is(err, io.EOF) || buf[n] < '!':
+			return n, nil
+		case err != nil:
+			return n, err
+		}
+	}
+	return n, nil
+}
+
+const fileEndLimit = 1
+
+// checkKeyFileEnd skips over additional newlines at the end of a key file.
+func checkKeyFileEnd(r io.ByteReader) error {
+	for idx := 0; ; idx++ {
+		b, err := r.ReadByte()
+		switch {
+		case errors.Is(err, io.EOF):
+			return nil
+		case err != nil:
+			return err
+		case b != '\n' && b != '\r':
+			return ErrInvalidPrivateKeyEnding
+		case idx > fileEndLimit:
+			return ErrInvalidPrivateKeyLen
+		}
+	}
+}
+
+func encodePrivateKey(pk *crypto.PrivateKeySECP256K1R) (string, error) {
+	privKeyRaw := pk.Bytes()
+	enc, err := formatting.EncodeWithChecksum(formatting.CB58, privKeyRaw)
+	if err != nil {
+		return "", err
+	}
+	return privKeyEncPfx + enc, nil
+}
+
+func decodePrivateKey(enc string) (*crypto.PrivateKeySECP256K1R, error) {
+	rawPk := strings.Replace(enc, privKeyEncPfx, "", 1)
+	skBytes, err := formatting.Decode(formatting.CB58, rawPk)
+	if err != nil {
+		return nil, err
+	}
+	rpk, err := keyFactory.ToPrivateKey(skBytes)
+	if err != nil {
+		return nil, err
+	}
+	privKey, ok := rpk.(*crypto.PrivateKeySECP256K1R)
+	if !ok {
+		return nil, ErrInvalidType
+	}
+	return privKey, nil
+}
+
+// Returns the private key.
+func (m *SoftKey) Key() *crypto.PrivateKeySECP256K1R {
+	return m.privKey
+}
+
+// Returns the private key in raw bytes.
+func (m *SoftKey) Raw() []byte {
+	return m.privKeyRaw
+}
+
+// Returns the private key encoded in CB58 and "PrivateKey-" prefix.
+func (m *SoftKey) Encode() string {
+	return m.privKeyEncoded
+}
+
+// Saves the private key to disk with hex encoding.
+func (m *SoftKey) Save(p string) error {
+	k := hex.EncodeToString(m.privKeyRaw)
```

Comment:
This will save the hex encoded version of the file. I think there is no agreement and what should be used, but the ewoq key is also in the `PrivateKey-...` format, I personally prefer to write the `privKeyEncoded` version instead. There was some discussion about this in the `subnet-cli` repo. Maybe @aaronbuchwald wants to add some comments?

Reply:
I don't really have strong feelings one way or the other.

Reply:
OK let's leave it at this, this is how subnet-cli has it too.